File: 6_tree/树状数组/经典题/二维/BIT2D-compress.go
// !注意速度不如一般的 RectangleSum, 谨慎使用 (2e5, 2200ms; 1e5, 1000ms)
// Add : 单点加
// Query : 区间和
// QueryPrefix : 前缀和

package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func main() {
	// https://judge.yosupo.jp/problem/point_add_rectangle_sum

	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	xs, ys, ws := make([]int, n), make([]int, n), make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &xs[i], &ys[i], &ws[i])
	}
	query := make([][4]int, q)
	for i := 0; i < q; i++ {
		var t int
		fmt.Fscan(in, &t)
		if t == 0 {
			var x, y, w int
			fmt.Fscan(in, &x, &y, &w)
			xs = append(xs, x)
			ys = append(ys, y)
			ws = append(ws, 0)
			query[i] = [4]int{-1, x, y, w}
		} else {
			var a, b, c, d int
			fmt.Fscan(in, &a, &b, &c, &d)
			query[i] = [4]int{a, c, b, d}
		}
	}

	tree := NewFenwickTree2DWithWeights(xs, ys, ws, true)
	for i := 0; i < q; i++ {
		a, b, c, d := query[i][0], query[i][1], query[i][2], query[i][3]
		if a == -1 {
			tree.Add(b, c, d)
		} else {
			fmt.Fprintln(out, tree.Query(a, b, c, d))
		}
	}
}

func rectangle_sum() {
	// https://judge.yosupo.jp/problem/rectangle_sum
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	xs, ys, ws := make([]int, n), make([]int, n), make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &xs[i], &ys[i], &ws[i])
	}
	tree := NewFenwickTree2DWithWeights(xs, ys, ws, true)
	for i := 0; i < q; i++ {
		var l, d, r, u int
		fmt.Fscan(in, &l, &d, &r, &u)
		fmt.Fprintln(out, tree.Query(l, r, d, u))
	}
}

func demo() {
	xs, ys := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
	tree := NewFenwickTree2D(xs, ys, false)
	tree.Add(1, 1, 91)
	fmt.Println(tree.Query(1, 2, 1, 2))
	fmt.Println(tree.QueryPrefix(1, 2))
}

type Able = int

// 需要是阿贝尔群(满足交换律)
func e() Able           { return 0 }
func op(a, b Able) Able { return a + b }
func inv(a Able) Able   { return -a }

type FenwickTree2D struct {
	n          int
	keyX       []int
	keyY       []int
	minX       int
	indptr     []int
	data       []Able
	discretize bool
	unit       Able
}

// discretize:
//  为 true 时对x维度二分离散化,然后用离散化后的值作为下标.
//  为 false 时不对x维度二分离散化,而是直接用x的值作为下标(所有x给一个偏移量minX),
//  x 维度数组长度为最大值减最小值.
func NewFenwickTree2D(xs, ys []int, discretize bool) *FenwickTree2D {
	res := &FenwickTree2D{discretize: discretize, unit: e()}
	ws := make([]Able, len(xs))
	for i := range ws {
		ws[i] = res.unit
	}
	res._build(xs, ys, ws)
	return res
}

// discretize:
//  为 true 时对x维度二分离散化,然后用离散化后的值作为下标.
//  为 false 时不对x维度二分离散化,而是直接用x的值作为下标(所有x给一个偏移量minX),
//  x 维度数组长度为最大值减最小值.
func NewFenwickTree2DWithWeights(xs, ys []int, ws []Able, discretize bool) *FenwickTree2D {
	res := &FenwickTree2D{discretize: discretize, unit: e()}
	res._build(xs, ys, ws)
	return res
}

// 点 (x,y) 的值加上 val.
func (fwt *FenwickTree2D) Add(x, y int, val Able) {
	i := fwt._xtoi(x)
	for ; i < fwt.n; i += ((i + 1) & -(i + 1)) {
		fwt._add(i, y, val)
	}
}

// [lx,rx) * [ly,ry)
func (t *FenwickTree2D) Query(lx, rx, ly, ry int) Able {
	pos, neg := t.unit, t.unit
	l, r := t._xtoi(lx)-1, t._xtoi(rx)-1
	for l < r {
		pos = op(pos, t._prodI(r, ly, ry))
		r -= ((r + 1) & -(r + 1))
	}
	for r < l {
		neg = op(neg, t._prodI(l, ly, ry))
		l -= ((l + 1) & -(l + 1))
	}
	return op(pos, inv(neg))
}

// [0,rx) * [0,ry)
func (t *FenwickTree2D) QueryPrefix(rx, ry int) Able {
	pos := t.unit
	r := t._xtoi(rx) - 1
	for r >= 0 {
		pos = op(pos, t._prefixProdI(r, ry))
		r -= ((r + 1) & -(r + 1))
	}
	return pos
}

func (t *FenwickTree2D) _add(i int, y int, val Able) {
	lid := t.indptr[i]
	n := t.indptr[i+1] - t.indptr[i]
	j := sort.SearchInts(t.keyY[lid:lid+n], y)
	for j < n {
		t.data[lid+j] = op(t.data[lid+j], val)
		j += ((j + 1) & -(j + 1))
	}
}

func (t *FenwickTree2D) _prodI(i int, ly, ry int) Able {
	pos, neg := t.unit, t.unit
	lid := t.indptr[i]
	n := t.indptr[i+1] - t.indptr[i]
	it := t.keyY[lid : lid+n]
	left := sort.SearchInts(it, ly) - 1
	right := sort.SearchInts(it, ry) - 1
	for left < right {
		pos = op(pos, t.data[lid+right])
		right -= ((right + 1) & -(right + 1))
	}
	for right < left {
		neg = op(neg, t.data[lid+left])
		left -= ((left + 1) & -(left + 1))
	}
	return op(pos, inv(neg))
}

func (t *FenwickTree2D) _prefixProdI(i int, ry int) Able {
	pos := t.unit
	lid := t.indptr[i]
	n := t.indptr[i+1] - t.indptr[i]
	it := t.keyY[lid : lid+n]
	R := sort.SearchInts(it, ry) - 1
	for R >= 0 {
		pos = op(pos, t.data[lid+R])
		R -= ((R + 1) & -(R + 1))
	}
	return pos
}

func (ft *FenwickTree2D) _build(X, Y []int, wt []Able) {
	if len(X) != len(Y) || len(X) != len(wt) {
		panic("Lengths of X, Y, and wt must be equal.")
	}

	if ft.discretize {
		ft.keyX = unique(X)
		ft.n = len(ft.keyX)
	} else {
		if len(X) > 0 {
			min_, max_ := 0, 0
			for _, x := range X {
				if x < min_ {
					min_ = x
				}
				if x > max_ {
					max_ = x
				}
			}
			ft.minX = min_
			ft.n = max_ - min_ + 1
		}
		ft.keyX = make([]int, ft.n)
		for i := 0; i < ft.n; i++ {
			ft.keyX[i] = ft.minX + i
		}
	}

	N := ft.n
	keyYRaw := make([][]int, N)
	datRaw := make([][]Able, N)
	indices := argSort(Y)

	for _, i := range indices {
		ix := ft._xtoi(X[i])
		y := Y[i]
		for ix < N {
			kY := keyYRaw[ix]
			if len(kY) == 0 || kY[len(kY)-1] < y {
				keyYRaw[ix] = append(keyYRaw[ix], y)
				datRaw[ix] = append(datRaw[ix], wt[i])
			} else {
				datRaw[ix][len(datRaw[ix])-1] = op(datRaw[ix][len(datRaw[ix])-1], wt[i])
			}
			ix += ((ix + 1) & -(ix + 1))
		}
	}

	ft.indptr = make([]int, N+1)
	for i := 0; i < N; i++ {
		ft.indptr[i+1] = ft.indptr[i] + len(keyYRaw[i])
	}
	ft.keyY = make([]int, ft.indptr[N])
	ft.data = make([]Able, ft.indptr[N])

	for i := 0; i < N; i++ {
		for j := 0; j < ft.indptr[i+1]-ft.indptr[i]; j++ {
			ft.keyY[ft.indptr[i]+j] = keyYRaw[i][j]
			ft.data[ft.indptr[i]+j] = datRaw[i][j]
		}
	}

	for i := 0; i < N; i++ {
		n := ft.indptr[i+1] - ft.indptr[i]
		for j := 0; j < n-1; j++ {
			k := j + ((j + 1) & -(j + 1))
			if k < n {
				ft.data[ft.indptr[i]+k] = op(ft.data[ft.indptr[i]+k], ft.data[ft.indptr[i]+j])
			}
		}
	}
}

func (ft *FenwickTree2D) _xtoi(x int) int {
	if ft.discretize {
		return sort.SearchInts(ft.keyX, x)
	}
	tmp := x - ft.minX
	if tmp < 0 {
		tmp = 0
	} else if tmp > ft.n {
		tmp = ft.n
	}
	return tmp
}

func unique(nums []int) (sorted []int) {
	set := make(map[int]struct{}, len(nums))
	for _, v := range nums {
		set[v] = struct{}{}
	}
	sorted = make([]int, 0, len(set))
	for k := range set {
		sorted = append(sorted, k)
	}
	sort.Ints(sorted)
	return
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func mins(nums ...int) int {
	res := nums[0]
	for _, num := range nums {
		if num < res {
			res = num
		}
	}
	return res
}

func maxs(nums ...int) int {
	res := nums[0]
	for _, num := range nums {
		if num > res {
			res = num
		}
	}
	return res
}

func argSort(nums []int) []int {
	order := make([]int, len(nums))
	for i := range order {
		order[i] = i
	}
	sort.Slice(order, func(i, j int) bool { return nums[order[i]] < nums[order[j]] })
	return order
}

File: 6_tree/树状数组/经典题/二维/BIT2D-compress.ts
/* eslint-disable max-len */
/* eslint-disable no-inner-declarations */
/* eslint-disable no-param-reassign */

// 二维树状数组 单点加 区间查询 (适用于值域很大的情况)
// Add : 单点加
// Query : 区间和
// QueryPrefix : 前缀和

interface IOptions<E> {
  xs: number[]
  ys: number[]
  ws?: E[]

  /**
   * 是否离散化x.
   * 为 true 时对x维度二分离散化,然后用离散化后的值作为下标.
   * 为 false 时不对x维度二分离散化,而是直接用x的值作为下标(内部给x一个偏移量minX),此时x维度数组长度为最大值减最小值.
   * 默认为 true.
   */
  discretizeX?: boolean

  e: () => E
  op: (a: E, b: E) => E
  inv: (a: E) => E
}

class BIT2DCompress<E> {
  private static _bisectLeft(arr: ArrayLike<number>, x: number, left: number, right: number): number {
    while (left <= right) {
      const mid = (left + right) >> 1
      if (arr[mid] < x) {
        left = mid + 1
      } else {
        right = mid - 1
      }
    }
    return left
  }

  private readonly _e: () => E
  private readonly _op: (a: E, b: E) => E
  private readonly _inv: (a: E) => E
  private readonly _discretizeX: boolean
  private _keyX!: number[]
  private _keyY!: number[]
  private _indptr!: Uint32Array
  private _data!: E[]
  private _n = 0
  private _minX = 0

  constructor(options: IOptions<E>) {
    let { xs, ys, e, op, inv, ws, discretizeX = true } = options
    this._e = e
    this._op = op
    this._inv = inv
    this._discretizeX = discretizeX
    if (!ws) {
      ws = Array(xs.length)
      for (let i = 0; i < xs.length; i++) ws[i] = e()
    }
    this._build(xs, ys, ws)
  }

  /**
   * 点 (x,y) 的值加上 val.
   */
  add(x: number, y: number, val: E): void {
    for (let i = this._xtoi(x); i < this._n; i += (i + 1) & -(i + 1)) {
      this._add(i, y, val)
    }
  }

  /**
   * [lx,rx) * [ly,ry) 的和.
   */
  query(lx: number, rx: number, ly: number, ry: number): E {
    if (lx >= rx || ly >= ry) return this._e()
    let pos = this._e()
    let neg = this._e()
    let l = this._xtoi(lx) - 1
    let r = this._xtoi(rx) - 1
    while (l < r) {
      pos = this._op(pos, this._prodI(r, ly, ry))
      r -= (r + 1) & -(r + 1)
    }
    while (r < l) {
      neg = this._op(neg, this._prodI(l, ly, ry))
      l -= (l + 1) & -(l + 1)
    }
    return this._op(pos, this._inv(neg))
  }

  /**
   * [0,rx) * [0,ry) 的和.
   */
  queryPrefix(rx: number, ry: number): E {
    let pos = this._e()
    let r = this._xtoi(rx) - 1
    while (r >= 0) {
      pos = this._op(pos, this._prefixProdI(r, ry))
      r -= (r + 1) & -(r + 1)
    }
    return pos
  }

  private _add(i: number, y: number, val: E): void {
    const lid = this._indptr[i]
    const n = this._indptr[i + 1] - this._indptr[i]
    let j = BIT2DCompress._bisectLeft(this._keyY, y, lid, lid + n - 1) - lid
    while (j < n) {
      this._data[lid + j] = this._op(this._data[lid + j], val)
      j += (j + 1) & -(j + 1)
    }
  }

  private _prodI(i: number, ly: number, ry: number): E {
    let pos = this._e()
    let neg = this._e()
    const lid = this._indptr[i]
    const n = this._indptr[i + 1] - this._indptr[i]
    let left = BIT2DCompress._bisectLeft(this._keyY, ly, lid, lid + n - 1) - lid - 1
    let right = BIT2DCompress._bisectLeft(this._keyY, ry, lid, lid + n - 1) - lid - 1
    while (left < right) {
      pos = this._op(pos, this._data[lid + right])
      right -= (right + 1) & -(right + 1)
    }
    while (right < left) {
      neg = this._op(neg, this._data[lid + left])
      left -= (left + 1) & -(left + 1)
    }
    return this._op(pos, this._inv(neg))
  }

  private _prefixProdI(i: number, ry: number): E {
    let pos = this._e()
    const lid = this._indptr[i]
    const n = this._indptr[i + 1] - this._indptr[i]
    let right = BIT2DCompress._bisectLeft(this._keyY, ry, lid, lid + n - 1) - lid - 1
    while (right >= 0) {
      pos = this._op(pos, this._data[lid + right])
      right -= (right + 1) & -(right + 1)
    }
    return pos
  }

  private _build(xs: number[], ys: number[], ws: E[]): void {
    if (xs.length !== ys.length || xs.length !== ws.length) {
      throw new Error('Lengths of X, Y, and wt must be equal.')
    }

    if (this._discretizeX) {
      this._keyX = [...new Set(xs)].sort((a, b) => a - b)
      this._n = this._keyX.length
    } else {
      if (xs.length > 0) {
        let min = xs[0]
        let max = xs[0]
        for (let i = 0; i < xs.length; i++) {
          min = Math.min(min, xs[i])
          max = Math.max(max, xs[i])
        }
        this._minX = min
        this._n = max - min + 1
      }
      this._keyX = Array(this._n)
      for (let i = 0; i < this._n; i++) {
        this._keyX[i] = this._minX + i
      }
    }

    const n = this._n
    const keyYRaw: number[][] = Array(n)
    const datRaw: E[][] = Array(n)
    for (let i = 0; i < n; i++) {
      keyYRaw[i] = []
      datRaw[i] = []
    }
    const indices = Array(ys.length)
    for (let i = 0; i < ys.length; i++) {
      indices[i] = i
    }
    indices.sort((a, b) => ys[a] - ys[b])

    for (let i = 0; i < indices.length; i++) {
      const v = indices[i]
      let ix = this._xtoi(xs[v])
      const y = ys[v]
      while (ix < n) {
        const ky = keyYRaw[ix]
        const tmp = datRaw[ix]
        if (ky.length === 0 || ky[ky.length - 1] < y) {
          keyYRaw[ix].push(y)
          tmp.push(ws[v])
        } else {
          tmp[tmp.length - 1] = this._op(tmp[tmp.length - 1], ws[v])
        }
        ix += (ix + 1) & -(ix + 1)
      }
    }

    this._indptr = new Uint32Array(n + 1)
    for (let i = 0; i < n; i++) {
      this._indptr[i + 1] = this._indptr[i] + keyYRaw[i].length
    }
    this._keyY = Array(this._indptr[n]).fill(0)
    this._data = Array(this._indptr[n])
    for (let i = 0; i < this._data.length; i++) {
      this._data[i] = this._e()
    }

    for (let i = 0; i < n; i++) {
      for (let j = 0; j < this._indptr[i + 1] - this._indptr[i]; j++) {
        this._keyY[this._indptr[i] + j] = keyYRaw[i][j]
        this._data[this._indptr[i] + j] = datRaw[i][j]
      }
    }

    for (let i = 0; i < n; i++) {
      const diff = this._indptr[i + 1] - this._indptr[i]
      for (let j = 0; j < diff - 1; j++) {
        const k = j + ((j + 1) & -(j + 1))
        if (k < diff) {
          this._data[this._indptr[i] + k] = this._op(this._data[this._indptr[i] + k], this._data[this._indptr[i] + j])
        }
      }
    }
  }

  private _xtoi(x: number): number {
    if (this._discretizeX) return BIT2DCompress._bisectLeft(this._keyX, x, 0, this._n - 1)
    const res = x - this._minX
    if (res < 0) return 0
    if (res >= this._n) return this._n
    return res
  }
}

if (require.main === module) {
  const tree = new BIT2DCompress({
    xs: [0, 0, 0, 1, 1, 1, 2, 2, 2],
    ys: [0, 1, 2, 0, 1, 2, 0, 1, 2],
    ws: [1, 2, 3, 4, 5, 6, 7, 8, 9],
    op: (a, b) => a + b,
    e: () => 0,
    inv: a => -a
  })
  console.log(tree.query(0, 3, 0, 3))

  // https://leetcode.cn/problems/range-sum-query-2d-mutable/submissions/
  class NumMatrix {
    private readonly matrix: number[][]
    private readonly ROW: number
    private readonly COL: number
    private readonly tree: BIT2DCompress<number>

    constructor(matrix: number[][]) {
      this.matrix = matrix
      this.ROW = matrix.length
      this.COL = matrix[0].length
      const xs = Array(this.ROW * this.COL)
      const ys = Array(this.ROW * this.COL)
      const ws = Array(this.ROW * this.COL)
      for (let r = 0; r < this.ROW; r++) {
        for (let c = 0; c < this.COL; c++) {
          xs[r * this.COL + c] = r
          ys[r * this.COL + c] = c
          ws[r * this.COL + c] = matrix[r][c]
        }
      }
      this.tree = new BIT2DCompress({
        xs,
        ys,
        ws,
        e: () => 0,
        op: (a, b) => a + b,
        inv: a => -a
      })
    }

    update(row: number, col: number, val: number): void {
      const pre = this.matrix[row][col]
      const diff = val - pre
      this.matrix[row][col] = val
      this.tree.add(row, col, diff)
    }

    sumRegion(row1: number, col1: number, row2: number, col2: number): number {
      return this.tree.query(row1, row2 + 1, col1, col2 + 1)
    }
  }
}

export { BIT2DCompress }
